Add a post about `rust-lang/rust`'s new LLM policy (#1897)
rust-lang/blog.rust-lang.org: content/inside-rust/rust-langrust-is-adopting-an-llm-policy.md

### Technical products no longer indicate effort

It used to be that if an open source project got a polished, well-tested, detailed PR,
that indicated that there was someone on the other end who had put time, effort, and understanding into the PR.
That influenced Rust's culture in several ways:
- We are generally reluctant to close PRs, since they represent someone else's hard work.
- Our process emphasizes incremental discussions, where PRs are allowed to change existing design if we discover new facts during creation or review.
- We treat PRs as an indication that someone is interested in joining our community and being mentored to work on future PRs.

With LLMs, none of these signals are reliable.
Polished PRs no longer indicate effort;
authors of polished PRs no longer necessarily understand their code—and in the case of autonomous agents, there is no longer someone on the other end at all;
and because it's become so much easier to write code, a polished PR no longer indicates that someone is likely to stick around for the long term.

### Making code easier to write causes review issues

> At time of writing, there are **1,281 open PRs** to rust-lang/rust.

Reviewer: ```suggestion
At the time of writing, there are **1,281 open PRs** to rust-lang/rust.
```

>
> Despite these differences, there are many values we all share:
>
> - Building a community of deep experts in our collective projects.
> - Building an inclusive community where all feel welcome and respected.

We want it to be possible to change the policy in the future—
the proposed [LLM committee](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/leadership-council/issues/308) is one possible way for that to happen;
the provisions in the policy about how to make amendments are another.

I do not think every rule in this policy is wholly good.
I *do* think that writing our rules down is better than not writing them down,
and that having a policy that everyone kinda dislikes pushes us to improve our governance structures.

## What does the policy say?

### General rules

The policy holds LLM-generated changes to a *higher* bar than human-authored changes, not a lower one:
LLM PRs are required to have tests, full stop, regardless of how hard that is, as well as various other restrictions;
LLMs cannot be used for soundness-critical changes unless the author is already a domain expert, and even then it's strongly discouraged.

> No one is required to read LLM output unless they choose to:

Reviewer 2: This sounds a bit misleading, it sounds like if you author a LLM PR, you are not required to read/review it, lol. Maybe "Other people" or "No one except the author"?

LLM output isn't allowed in public docs, PR descriptions, and comments unless it's clearly marked;
reviewers aren't required to look at LLM PRs if they don't want to.

> LLMs are not allowed to become an implicit requirement for contributing to the Rust project:

Reviewer 2: While I hope that we will have this either in the CoC or a Project-wide policy soon, this is a bit misleading, because 1040 is not about contributing to the Rust Project, but about contributing to `rust-lang/rust`.

Author: edited it a bit, better?

Reviewer 2: Yup, thanks!

Some parts of the policy are unenforceable.
This is not a bug.
The goal is *not* to catch every violation, but to create a clear bright-line rule:
Disclosure is required for all public LLM text, unless specifically exempted by the policy.
This allows moderators to identify violations based on *actions*, not on intent, and only consider intent when deciding how to respond.

## How does this affect contributors?

### Issue reporters

You must disclose your use of LLMs.
You must clearly quote and indicate which parts of your report were LLM-generated;
the "no LLM-generated comments" rule applies to you too.

### Authors who use LLMs

I have written a list of guidelines you should follow if you use an LLM to create a PR to `rust-lang/rust`.
You can avoid thinking about them, or indeed reading the list at all, if you follow this simple guideline from the policy:

> It's fine to use LLMs to answer questions, analyze, distill, refine, check, suggest, review. But not to **create**.

See the ["Allowed" section of the policy](https://forge.rust-lang.org/policies/llm-usage.html) for a full list of what's meant by that.
See [rustc-dev-guide][llm-writing] for the full list of guidelines.

> [llm-writing]: https://rustc-dev-guide.rust-lang.org/llm-guidance/writing.html

Reviewer 2: Just noting that this is 404 atm (probably should be, until 1040 lands).

### Reviewers and moderators

#### General review

You are allowed to close PRs that don't follow the policy, no questions asked.
Please point the author to [#llm-mentoring] at the same time.
See [the dev guide][llm-reviewing] for exact circumstances and suggested wording.

You are not responsible for determining whether a PR is LLM-generated;
that responsibility lies with the author.
Style is not evidence; please do not accuse people of using an LLM.

> We will add a PR template that asks authors whether they use an LLM so that this rarely comes up.

Reviewer 3: This raises the question of: How should reviewers/moderators proceed when faced with a PR that they think is using LLM, but doesn't disclose usage, or claims to not be using LLM?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1897/f35e1174d2c15be7b692863b874a8a0046722081..471230d2e572c00de9fda8c5c7c207ac152e0040)*

Author: i think *that’s* the point at which a moderator should be involved. one of my goals with the policy was to make the vast majority of PRs handleable by reviewers, i still count it as success if we have to call in mods once a while.

Reviewer 3: In that case, maybe the "Reviewers and moderators" section should be changed to just "Reviewers"?

And maybe spell it out explicitly that it's a moderator's job to handle these cases?

Author: good call, done

If you are not in one of those groups, you don't have to change anything about how you work.

## Why was this policy created?

While the Rust project is a collection of technical artifacts,
it is also a *community* of people who work together to build, maintain, and extend those artifacts.
When we talk about "contributing to the Rust project", we partly mean work on those artifacts,
but we also mean joining that community and collaborating with the people already there.

Even before this policy was created, people were using LLMs to contribute to rust-lang/rust.
Some of those uses respected our community:
translating messages to English so people could draft them in their native language;
finding poor diagnostics for code snippets that new contributors to Rust might write;
analyzing RFCs to see if they were missing a discussion of other parts of the language that could affect the design.
Some of those uses, sometimes unintentionally, did not.

I've seen LLMs causing three main issues for our community:
1. Polished technical products no longer indicate effort and understanding.
2. Making code easier to write exacerbates our existing issues with review bandwidth.

> 3. People blindly copy-pasting to and from an LLM is a waste of our time.

Reviewer 4: Small note: mindlessly is probably a better word than blindly, to avoid perpetuating the equation of the two.

Also applies to the later section titled this.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1897/f35e1174d2c15be7b692863b874a8a0046722081..a2f043ab6d8fcc67fb3829d048c17735fc516f8a)*

Reviewer 4: (also no idea why GitHub decided to include 23 extra lines on this comment, but I'm specifically talking about the last point)

Author: i've changed this to "mechanically" to try and get at the actual problem this causes.

> 3. People mechanically copy-pasting to and from an LLM is a waste of our time.

Reviewer 5: ```suggestion
3. People mechanically copy-pasting to and from an LLM is a waste of our time and a breach of trust with the contributor
```

I would like to point out the breach of trust issue somewhere, which is the fundamental problem underneath these points. Unsure, though, where the "right" spot is.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1897/f35e1174d2c15be7b692863b874a8a0046722081..324928dda71b0820c344c5079208e82374124e34)*

Author: seems good. i don't think this overview is quite the right place, but i mention it under the expanded "# Mechanical copy-pasting" section.

As time went on, these issues grew and grew, until we had to create dedicated channels and moderation policy for how to deal with them.
However, those channels work against our goals of being transparent and welcoming, because new contributors have no idea what the rules are.

The new policy formalizes those rules publicly, so that new contributors know how to join our community without getting their PRs closed for reasons they don't understand,

> and so that existing reviewers can easily point to the rules as an actionable reason when closing PRs that don't follow them.

Reviewer 5: ... and at the same time protect us from uncontrolled LLM generated content

(or something like that)

Author: i added a section on mental models a little further down, underneath "General rules".

- We treat PRs as an indication that someone is interested in joining our community and being mentored to work on future PRs.

With LLMs, none of these signals are reliable.
Polished PRs no longer indicate effort;
authors of polished PRs no longer necessarily understand their code—and in the case of autonomous agents, there is no longer someone on the other end at all;
and because it's become so much easier to write code, a polished PR no longer indicates that someone is likely to stick around for the long term.

### Making code easier to write causes review issues

At the time of writing, there are **1,281 open PRs** to rust-lang/rust.
This represents a staggering amount of time invested by both authors and reviewers.
We have long had the problem that there are more people who want to write code than people willing to review it.
With the advent of LLMs, this problem only gets worse.

Most of the work of reviewing is not simply catching bugs.
A great deal of it is deciding *whether this direction is a good approach*, whether the PR is a good idea at all.

> In other words, [reviewing is made of decisions](https://siderea.dreamwidth.org/1219758.html).

Reviewer 5: ```suggestion
In other words, [reviewing is made of decisions](https://web.archive.org/web/20260213080731/https://siderea.dreamwidth.org/1219758.html).
```

just to be on the side, would it make sense using an archived version of that page?

Why not ban LLMs altogether, or allow any use of LLMs that we think are pro-social?
Because Rust governance doesn't work that way.
We do not have a benevolent dictator who can say
["No LLM-generated content, whether it be code or prose."](https://ziglang.org/code-of-conduct/#strict-no-llm-no-ai-policy)
or ["AI is a tool, just like other tools we use"](https://lore.kernel.org/linux-media/CAHk-=wi4zC+Ze8e+p3tMv8TtG_80KzsZ1syL9anBtmEh5Z40vg@mail.gmail.com/).

Rust operates by consensus. As the policy says:

> There is not a consensus within the Rust project—and likely never will be—about when/how/where it is acceptable to use AI-based tools.
> Many members of the Rust project and community find value in AI;
> many others feel that its negative impact on society and the climate are severe enough that no use is acceptable.
> Still others are working out their opinion.
>
> Despite these differences, there are many values we all share:
>
> - Building a community of deep experts in our collective projects.
> - Building an inclusive community where all feel welcome and respected.

We want it to be possible to change the policy in the future—

> the proposed [LLM committee](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/leadership-council/issues/308) is one possible way for that to happen;

Reviewer 4: I'm not 100% sure sending random readers at R-L.org to this issue is a good idea, but I also don't know if it's a net negative or positive to mention the committee here. To me, it seems like the best approach would be to mention that the policy adds provisions to make amendments, but the community and specifically the Council are working on other options too.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1897/f35e1174d2c15be7b692863b874a8a0046722081..8a654f2d7a3de51df289bfb474780c78d5bb4f30)*

Author: @oli-obk it's your issue and you'd have to moderate it. thoughts? opinions?

Reviewer 4: I guess part of my motivation also is that I don't particularly expect the comments to be helpful, and mostly just a bunch of people misunderstanding what the proposal actually is. It doesn't help that the proposal itself has been changed a bunch of times. It just feels like the kind of issue that isn't going to be relevant outside of the project until it's actually accepted, at which point people can offer better feedback.

Reviewer 6: 🤷 council issues aren't really useful for discussions anyway, not sure the link will give us more noise on it. Can always lock it

Author: i added the link back; i think it clarifies that this is happening in public and not just in backrooms.

> The leadership council is also considering creating a sub-team that would handle LLM policy so we have fewer "nightmare" 30-person FCPs.

Reviewer 7: I assume the target audience of this post goes well beyond the project, so maybe we should avoid jargon like "FCP"?

Author: nice, done

I do not think every rule in this policy is wholly good.
I *do* think that writing our rules down is better than not writing them down,
and that having a policy that everyone kinda dislikes pushes us to improve our governance structures.

## What does the policy say?

### General rules

The policy holds LLM-generated changes to a *higher* bar than human-authored changes, not a lower one:
LLM PRs are required to have tests, full stop, regardless of how hard that is, as well as various other restrictions;
LLMs cannot be used for soundness-critical changes unless the author is already a domain expert, and even then it's strongly discouraged.

No one except the author is required to read LLM output unless they choose to:
LLM output isn't allowed in public docs, PR descriptions, and comments unless it's clearly marked;
reviewers aren't required to look at LLM PRs if they don't want to.

LLMs are not allowed to become an implicit requirement for contributing to `rust-lang/rust`:
policies must be written first for humans, and only summarized for machines;

> LLM PRs without disclosure can be closed for that reason alone.

Reviewer 7: This last line here seems to fit better to the previous paragraph ("No one except the author is required to read LLM output") that this one ("not allowed to become an implicit requirement")?

Author: yes, makes sense. replaced it with "LLM reviews cannot substitute for human review or self-review."

Reviewer 7: I don't quite see the causal link from "not allowed to become an implicit requirement" to "cannot substitute for human review or self-review"?

Author: so, one story i've heard about repos that use LLMs heavily for writing code is that they also use LLMs heavily for *reviewing* code. Eventually someone has the bright idea that they could merge code much faster if they *only* required LLM review, and now anyone who *doesn't* use LLMs is at a disadvantage because they have to wait for another person while their colleagues are ~self-approving. at that point the LLM is a de-facto requirement even if no one's written it down.

(i know @nnethercote has talked about this a bit in the abstract, but i've also heard real stories about it from people working on teams at large corporations.)

Reviewer 7: I see... "non-LLM-users being disadvantaged" is not the same as "LLM being required" so I didn't quite make that connection.

Author: i've changed the section heading to "People must not be disadvantaged if they do not use an LLM for contributing to `rust-lang/rust`".

You can avoid thinking about them, or indeed reading the list at all, if you follow this simple guideline from the policy:

> It's fine to use LLMs to answer questions, analyze, distill, refine, check, suggest, review. But not to **create**.

See the ["Allowed" section of the policy](https://forge.rust-lang.org/policies/llm-usage.html) for a full list of what's meant by that.
See [rustc-dev-guide][llm-writing] for the full list of guidelines.

[llm-writing]: https://rustc-dev-guide.rust-lang.org/llm-guidance/writing.html

### Reviewers

#### General review

You are allowed to close PRs that don't follow the policy, no questions asked.
Please point the author to [#llm-mentoring] at the same time.
See [the dev guide][llm-reviewing] for exact circumstances and suggested wording.

You are not responsible for determining whether a PR is LLM-generated;
that responsibility lies with the author.
We will add a PR template that asks authors whether they use an LLM so that this rarely comes up.

If an author claims their code is not LLM-generated, but you're still not sure, please report the PR *privately* to moderation.
Style is not evidence; please do not accuse people of using an LLM.

> Reporting is not intended to be a penalty; the mod team is interested in seeing non-violations as well as violations

Reviewer 4: ```suggestion
Reporting is not intended to be a penalty; the mod team is interested in seeing non-violations as well as violations.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1897/f35e1174d2c15be7b692863b874a8a0046722081..a2f043ab6d8fcc67fb3829d048c17735fc516f8a)*

> many others feel that its negative impact on society and the climate are severe enough that no use is acceptable.
> Still others are working out their opinion.
>
> Despite these differences, there are many values we all share:
>
> - Building a community of deep experts in our collective projects.
> - Building an inclusive community where all feel welcome and respected.

We want it to be possible to change the policy in the future.
The policy has several provisions that make it easier to change than it was to originally adopt.
The leadership council is also considering creating a sub-team that would handle LLM policy so we have fewer "nightmare" 30-person approval requirements.

I do not think every rule in this policy is wholly good.
I *do* think that writing our rules down is better than not writing them down,
and that having a policy that everyone kinda dislikes pushes us to improve our governance structures.

## What does the policy say?

### General rules

The policy holds LLM-generated changes to a *higher* bar than human-authored changes, not a lower one:
LLM PRs are required to have tests, full stop, regardless of how hard that is, as well as various other restrictions;

> LLMs cannot be used for soundness-critical changes unless the author is already a domain expert, and even then it's strongly discouraged.

Reviewer 8: This feels like a strange place to start. It makes it sound like LLM-generated changes are basically accepted, when really they're not (except for the experimental stuff, which is off to the side).

I think the "analysis/review/distill yes; create no (with some strict exceptions)" framing is the best starting point. That's mentioned below; it could be moved up here.

Author: excellent idea, done.

No one except the author is required to read LLM output unless they choose to:
LLM output isn't allowed in public docs, PR descriptions, and comments unless it's clearly marked;
reviewers aren't required to look at LLM PRs if they don't want to.

LLMs are not allowed to become an implicit requirement for contributing to `rust-lang/rust`:
policies must be written first for humans, and only summarized for machines;
LLM reviews cannot substitute for human review or self-review.

> LLMs are allowed for your own personal use without disclosure, as long as you do not post LLM output that you expect us to read or review.

Reviewer 8: I think the use of the words "use" and "usage" are inconsistent, here and in several places elsewhere. E.g. compare this sentence with the later one "You must disclose LLM usage". I think "use" sometimes means "generate/author" and sometimes means "analyze/review".

I've highlighted a couple of additional case below, but I think you need to carefully go through every example of "use"/"usage" and clarify exactly which kind(s) of use is being meant. That's a pain, I know, but otherwise you can cherry-pick sentences from this that are quite misleading :(

Author: this is a really good point, thank you. i've gone through and tried to clarify this. i believe the only remaining places i say "use" are when i mean "any kind of use, not just particular ones", but i'd appreciate a double-check on that.

### Moderation rules

**You must disclose LLM usage.**
You can choose not to use an LLM, or you can choose to use it and disclose your use.
You may not hide that you used an LLM.

**Harassment is not allowed.**
You may not harass people for using an LLM, regardless of whether or not their use is banned by the policy.
You must follow the [Code of Conduct] at all times when interacting with the Rust project.

[Code of Conduct]: https://rust-lang.org/policies/code-of-conduct/

See [the policy itself][forge-page] for more information.

Some parts of the policy are unenforceable.
This is not a bug.
The goal is *not* to catch every violation, but to create a clear bright-line rule:
Disclosure is required for all public LLM text, unless specifically exempted by the policy.
This allows moderators to identify violations based on *actions*, not on intent, and only consider intent when deciding how to respond.

## How does this affect contributors?

### Issue reporters

You must disclose your use of LLMs.
You must clearly quote and indicate which parts of your report were LLM-generated;
the "no LLM-generated comments" rule applies to you too.

### Authors who use LLMs

> I have written a list of guidelines you should follow if you use an LLM to create a PR to `rust-lang/rust`.

Reviewer 8: As above: "use an LLM to create" is unclear. I think you mean the LLM is authoring the code, but "use" could be interpreted to cover analysis/review/etc.

You can avoid thinking about them, or indeed reading the list at all, if you follow this simple guideline from the policy:

> It's fine to use LLMs to answer questions, analyze, distill, refine, check, suggest, review. But not to **create**.

See the ["Allowed" section of the policy](https://forge.rust-lang.org/policies/llm-usage.html) for a full list of what's meant by that.
See [rustc-dev-guide][llm-writing] for the full list of guidelines.

[llm-writing]: https://rustc-dev-guide.rust-lang.org/llm-guidance/writing.html

### Reviewers

#### General review

You are allowed to close PRs that don't follow the policy, no questions asked.
Please point the author to [#llm-mentoring] at the same time.
See [the dev guide][llm-reviewing] for exact circumstances and suggested wording.

You are not responsible for determining whether a PR is LLM-generated;
that responsibility lies with the author.

> We will add a PR template that asks authors whether they use an LLM so that this rarely comes up.

Reviewer 8: Another ambiguous "use".

> If an author claims their code is not LLM-generated, but you're still not sure, please report the PR *privately* to moderation.

Reviewer 8: "LLM-generated" is clearer, that's good.

The policy affects the following groups of people:
- People who review or moderate PRs on `rust-lang/rust`.
- People who author PRs with LLM-generated code on `rust-lang/rust`.
- People who discover issues using LLMs and post them on `rust-lang/rust`.
- People who write issues or comments that directly quote an LLM on `rust-lang/rust`.

If you are not in one of those groups, you don't have to change anything about how you work.

## Why was this policy created?

While the Rust project is a collection of technical artifacts,
it is also a *community* of people who work together to build, maintain, and extend those artifacts.
When we talk about "contributing to the Rust project", we partly mean work on those artifacts,
but we also mean joining that community and collaborating with the people already there.

Even before this policy was created, people were using LLMs to contribute to rust-lang/rust.
Some of those uses respected our community:
translating messages to English so people could draft them in their native language;
finding poor diagnostics for code snippets that new contributors to Rust might write;

> analyzing RFCs to see if they were missing a discussion of other parts of the language that could affect the design.

Reviewer 9: Would this be easier to read with bullet points at the start of the last few lines?

Author: i like this as-is and don't want to change it.

and so that existing reviewers can easily point to the rules as an actionable reason when closing PRs that don't follow them.

### Technical products no longer indicate effort

It used to be that if an open source project got a polished, well-tested, detailed PR,
that indicated that there was someone on the other end who had put time, effort, and understanding into the PR.
That influenced Rust's culture in several ways:
- We are generally reluctant to close PRs, since they represent someone else's hard work.
- Our process emphasizes incremental discussions, where PRs are allowed to change existing design if we discover new facts during creation or review.
- We treat PRs as an indication that someone is interested in joining our community and being mentored to work on future PRs.

With LLMs, none of these signals are reliable.
Polished PRs no longer indicate effort;
authors of polished PRs no longer necessarily understand their code—and in the case of autonomous agents, there is no longer someone on the other end at all;

> and because it's become so much easier to write code, a polished PR no longer indicates that someone is likely to stick around for the long term.

Reviewer 9: Same question here, is this more readable as a bulleted list?

"Shotgunning" PRs at reviewers incurs a high mental cost for them.
I think most authors of LLM PRs believe that they are sincerely helping, but from our perspective,
the code itself is the smallest and in some ways least important part of the change.
We care much more about authors *understanding* what the code does, *planning* how it will change in the future, and *deciding* what it should look like.
The code itself cannot help with any of those.

### Mechanically copy-pasting LLM output is a waste of time

We will often get people who respond to review comments by copy-pasting them into their LLM, then copy-pasting its response back onto GitHub.
Bluntly: **this is a waste of everyone's time**.
If we wanted an LLM's opinion, we could have asked it ourselves.
We want to hear *your* thoughts, not a machine's.

Moreover, this is a breach of trust between the reviewer and the author.
Our assumption when we review is that we're talking to a real person who wants to do their best work.
Pasting LLM text creates suspicion: does the author actually care? Is there a person here at all?

### So why a policy?

Before this policy, we had a "wild west" approach to moderation.

> We had dozens of LLM PRs; no disclosure rules; people trying to add MIR optimizations as their first PR;

Reviewer 9: I thought of expanding the acronym, but it doesn't actually mean much out of context. Linking to the reference and adding "risky" might help here.

```suggestion
We had dozens of LLM PRs; no disclosure rules; people trying to add risky [MIR](https://rustc-dev-guide.rust-lang.org/mir/index.html) optimizations as their first PR;
```

> There is not a consensus within the Rust project—and likely never will be—about when/how/where it is acceptable to use AI-based tools.
> Many members of the Rust project and community find value in AI;
> many others feel that its negative impact on society and the climate are severe enough that no use is acceptable.
> Still others are working out their opinion.
>
> Despite these differences, there are many values we all share:
>
> - Building a community of deep experts in our collective projects.
> - Building an inclusive community where all feel welcome and respected.

We want it to be possible to change the policy in the future.
The policy has several provisions that make it easier to change than it was to originally adopt.
The leadership council is also considering creating a sub-team that would handle LLM policy so we have fewer "nightmare" 30-person approval requirements.

I do not think every rule in this policy is wholly good.
I *do* think that writing our rules down is better than not writing them down,
and that having a policy that everyone kinda dislikes pushes us to improve our governance structures.

## What does the policy say?

> The shortest summary of the policy is this:

Reviewer 9: It wasn't clear to me that this is a quote from the policy (rather than something written just for this blog post)

```suggestion
The shortest summary in the policy is this:
```

Author: is this a useful distinction? will it matter to anyone reading the post?

Reviewer 9: It impacts how authoritative the quote is, but I see what you mean, maybe quoting it is enough of a hint.

Author: yeah, fair enough. clarified the wording.

> It's fine to use LLMs to answer questions, analyze, distill, refine, check, suggest, review. But not to **create**.

Uses in the first category are allowed, sometimes requiring disclosure.
Uses in the second category are heavily restricted.

### General rules

No one except the author is required to read LLM output unless they choose to:
LLM output isn't allowed in public docs, PR descriptions, and comments unless it's clearly marked;
reviewers aren't required to look at LLM PRs if they don't want to.

> People must not be disadvantaged if they do not use an LLM for contributing to `rust-lang/rust`:

Reviewer 9: I find triple negative sentences hard to parse, but I can't think of a neat re-wording.

Here are some possible clarifications:
```suggestion
People must not be disadvantaged if they do not use an LLM for contributing to `rust-lang/rust`:
LLMs can't be required for effective participation (either formally or in practice);
```

```suggestion
People must not be disadvantaged if they do not use an LLM for contributing to `rust-lang/rust`:
participation can't require an LLM, even as a soft requirement for efficiency;
```

Author: there are already two sentences below clarifying what this means. do you think they're sufficient?

Reviewer 9: I think they're good examples of some tasks that are impacted, but I still found it hard to work out "why" (the overall goal).

I found [this part of the policy wording](https://github.com/rust-lang/rust-forge/pull/1040/changes#diff-2d95c811a5fd15552f76562ec278135a5cc9ac9579df3a8ba53e6b7de8d1d855R250) more readable:
```suggestion
Avoid making LLMs a requirement to contribute to `rust-lang/rust`:
```

policies must be written first for humans, and only summarized for machines;
LLM reviews cannot substitute for human review or self-review.

You are allowed to generate LLM content that only you see, without disclosure,
as long as you do not post it anywhere that you expect us to read or review.

Disclosure is required for machine translation, "trivial" changes, discovering bugs, and reviewing other people's work using an LLM.

> We welcome messages posted in your native language; English translation is not required to contribute.

Reviewer 9: "What about disabled people?" is a question that comes up a lot in these discussions (sometimes from people without previous experience of disability advocacy).

So this might be a useful point to mention machine voice transcription and similar accessibility tools. For a few people, that's the only way they can communicate in text. So it doesn't fall neatly into the "only you can see" exception.

Transcription tools used to be better trained, and quite separate from generative LLMs. But unfortunately a lot of products now include LLMs as a mandatory part of the pipeline, or use LLMs as their first stage.

Author: I do not want to create new policy in this blog post.

Reviewer 9: I assumed that accessibility tools were covered by the "machine translation" parts of the policy:
> - Using machine-translation (e.g. Google Translate) from your native language without posting your original message is allowed but discouraged.
  Doing so can introduce new miscommunications that weren't there originally, and prevents someone who speaks the language from providing a better translation.
>     - ℹ️ Posting both your original message and the translated version is always ok, but you must still disclose that machine-translation was used.

If they're not, then I agree we shouldn't change policy here.

Author: I wasn't thinking about accessibility tools when I wrote that paragraph. I wouldn't want to describe accessibility as "allowed but discouraged". so I think it's better to leave them out of the post.

There are very strict guidelines on LLM-generated code changes:
> Pre-arranged, non-critical, high-quality, well-tested, and well-reviewed code changes that are originally created by an LLM are allowed, **with disclosure**.

The policy holds LLM-generated changes to a *higher* bar than human-authored changes, not a lower one:
LLM PRs are required to have tests, full stop, regardless of how hard that is, as well as various other restrictions;
LLMs must not generate soundness-critical changes unless the author is already a domain expert, and even then it's strongly discouraged.

In general, the policy focuses on *understanding*, helping to ensure that we have a mental model of our code, not just artifacts that mechanically do the right thing.
Zig links [Profession by Isaac Asimov][asimov] as the motivation for their policy, and I believe that's the right motivation.
No programmer tapes.

[asimov]: https://web.archive.org/web/20201109034130/https://www.abelard.org/asimov.php

### Moderation rules

**You must disclose LLM-generated content.**
You can choose to not post LLM content, or you can choose to post it and disclose its origin.

> You may not hide that you used an LLM.

Author: this last sentence and one place below are the last remaining spots that say "use" when they mean something more specific. i'm not sure what to replace this with. i *think* the sentences above clarify it in context, but taken on its own it's still misleading.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1897/f35e1174d2c15be7b692863b874a8a0046722081..5f45cca8069d88a89c409229cf64366bfe78b657)*

Reviewer 10: "Do not hide how LLM involvement"?

**Harassment is not allowed.**
You may not harass people for using an LLM, regardless of whether or not their use is banned by the policy.
You must follow the [Code of Conduct] at all times when interacting with the Rust project.

[Code of Conduct]: https://rust-lang.org/policies/code-of-conduct/

See [the policy itself][forge-page] for more information.

Some parts of the policy are unenforceable.
This is not a bug.
The goal is *not* to catch every violation, but to create a clear bright-line rule:
Disclosure is required for all public LLM text, unless specifically exempted by the policy.
This allows moderators to identify violations based on *actions*, not on intent, and only consider intent when deciding how to respond.

## How does this affect contributors?

### Issue reporters

You must disclose your use of LLMs.
You must tell us if you found an issue using an LLM.
You must clearly quote and indicate which parts of your report were LLM-generated;

> the "no LLM-generated comments" rule applies to you too.

Author: this first sentence is also misleading when taken on its own instead of in context. i'm not sure what to replace it with.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1897/f35e1174d2c15be7b692863b874a8a0046722081..5f45cca8069d88a89c409229cf64366bfe78b657)*

Author: took my best shot at removing this but i'm not super happy with it.

+++
path = "inside-rust/9999/12/31/rust-langrust-is-adopting-an-llm-policy"
title = "rust-lang/rust is adopting an LLM policy"
authors = ["Jynn Nelson"]
+++

> Recently, [five teams in the Rust project][read-only FCP link] adopted a [policy][forge-page] that I originally authored,

Reviewer 8: The forge-page link is currently broken. Not sure if that is expected, i.e. whether it will be populated before this post is published.

Author: that’s expected. it’ll stop being a 404 once https://github.com/rust-lang/rust-forge/pull/1040 is merged.

governing how [Large Language Models][llms] can be used when contributing to the `rust-lang/rust` monorepo.
Notably, the new policy is *not* an official stance on LLMs, and does *not* apply everywhere in the Rust project.
I wrote it for a very specific purpose, described below.

[llms]: https://en.wikipedia.org/wiki/Large_language_model

> This post talks about why we ratified that policy, what it says, and how this will affect contributors.

Reviewer 11: ```suggestion
This post talks about why we approved that policy, what it says, and how this will affect contributors.
```
simpler for non-native English speakers. If you strongly prefer ratified. Feel free to resolve this comment 👍

The policy affects the following groups of people:
- People who review or moderate PRs on `rust-lang/rust`.
- People who author PRs with LLM-generated code on `rust-lang/rust`.
- People who discover issues using LLMs and post them on `rust-lang/rust`.
- People who write issues or comments that directly quote an LLM on `rust-lang/rust`.

If you are not in one of those groups, you don't have to change anything about how you work.

## Why was this policy created?

While the Rust project is a collection of technical artifacts,
it is also a *community* of people who work together to build, maintain, and extend those artifacts.
When we talk about "contributing to the Rust project", we partly mean work on those artifacts,
but we also mean joining that community and collaborating with the people already there.

> Even before this policy was created, people were using LLMs to contribute to rust-lang/rust.

Reviewer 11: ```suggestion
Even before this policy was created, people were using LLMs to contribute to `rust-lang/rust`.
```

### Technical products no longer indicate effort

It used to be that if an open source project got a polished, well-tested, detailed PR,
that indicated that there was someone on the other end who had put time, effort, and understanding into the PR.
That influenced Rust's culture in several ways:
- We are generally reluctant to close PRs, since they represent someone else's hard work.
- Our process emphasizes incremental discussions, where PRs are allowed to change existing design if we discover new facts during creation or review.
- We treat PRs as an indication that someone is interested in joining our community and being mentored to work on future PRs.

With LLMs, none of these signals are reliable.
Polished PRs no longer indicate effort;
authors of polished PRs no longer necessarily understand their code—and in the case of autonomous agents, there is no longer someone on the other end at all;
and because it's become so much easier to write code, a polished PR no longer indicates that someone is likely to stick around for the long term.

### Making code easier to write causes review issues

> At the time of writing, there are **1,281 open PRs** to rust-lang/rust.

Reviewer 11: ```suggestion
At the time of writing, there are **1,281 open PRs** to `rust-lang/rust`.
```

> Still others are working out their opinion.
>
> Despite these differences, there are many values we all share:
>
> - Building a community of deep experts in our collective projects.
> - Building an inclusive community where all feel welcome and respected.

We want it to be possible to change the policy in the future.
The policy has [several provisions][modification-rules] that make it easier to change than it was to originally adopt.
The leadership council is also [considering creating a sub-team][llm-committee] that would handle LLM policy so we have fewer "nightmare" 30-person approval requirements.

[modification-rules]: https://forge.rust-lang.org/policies/llm-usage.html#conditions-for-modification-or-dissolution
[llm-committee]: https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/leadership-council/issues/308

I do not think every rule in this policy is wholly good.
I *do* think that writing our rules down is better than not writing them down,
and that having a policy that everyone kinda dislikes pushes us to improve our governance structures.

## What does the policy say?

The policy summarizes itself this way:

> > It's fine to use LLMs to answer questions, analyze, distill, refine, check, suggest, review. But not to **create**.

Reviewer 11: ```suggestion
> It's fine to use LLMs to answer questions, analyze, distill, refine, check, suggest, review.
> Pre-arranged, non-critical, high-quality, well-tested, and well-reviewed code changes that are originally created by an LLM are allowed, with disclosure.
```

from the current version of this line, it seems that "creating" with LLMs is forbidden.
Instead later you say it's restricted.
Consider rewording to improve clarity, either by specifying immediately the full rule or by rewording differently.

Author: i do immediately specify the full rule on the line below:

> Uses in the first category are allowed, sometimes requiring disclosure.
Uses in the second category are heavily restricted.

Reviewer 11: yes, exactly. I was confused because I felt like the two lines contradict each other

Reviewer 9: "Not fine" has a range of meanings from "tolerable" to "problematic". It can mean "forbidden", but that's rare. So I don't think there's much of a contradiction here.

(I like the changes to these paragraphs over the weekend, they make more sense to me now.)

Reviewer 5: FWIW I personally prefer the current version because it's unambiguous.

Uses in the first category are allowed, sometimes requiring disclosure.

> Uses in the second category are heavily restricted.

Reviewer 8: Great introduction to this section!

### General rules

No one except the author is required to read LLM output unless they choose to:

> LLM output isn't allowed in public docs, PR descriptions, and comments unless it's clearly marked;

Reviewer 12: It's a bit unclear what `comments` refers to here, but the policy seem to cover both GitHub comments and non-trivial code / doc comments, which is notable.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1897/a1312e164a99c339ca287820d4ebbec37e5a1a89..cea6988ab6a3d3db37431fde264bfd8d1d9c9307)*

Reviewer 9: I assumed it was PR descriptions and PR comments, given the context

Reviewer 4: Yeah, but it is technically ambiguous and probably should be specified as GitHub comments.

Reviewer 10: ```suggestion
LLM output isn't allowed in public docs, PR descriptions, or Github comments unless it's clearly marked;
```

[Code of Conduct]: https://rust-lang.org/policies/code-of-conduct/

See [the policy itself][forge-page] for more information.

Some parts of the policy are unenforceable.
This is not a bug.
The goal is *not* to catch every violation, but to create a clear bright-line rule:
Disclosure is required for all public LLM text, unless specifically exempted by the policy.
This allows moderators to identify violations based on *actions*, not on intent, and only consider intent when deciding how to respond.

## How does this affect contributors?

### Issue reporters

You must disclose any LLM involvement in discovering or reporting issues.
You must tell us if you found an issue using an LLM.
You must clearly quote and indicate which parts of your report were LLM-generated;
the "no LLM-generated comments" rule applies to you too.

### Authors who post LLM-generated code

I have written a list of guidelines you should follow if you make a PR to `rust-lang/rust` with LLM-generated code.
You can avoid thinking about them, or indeed reading the list at all, if you follow this simple guideline from the policy:

> > It's fine to use LLMs to answer questions, analyze, distill, refine, check, suggest, review. But not to **create**.

Reviewer 11: I find it weird to repeat this line again. The reader already got this point imo.

Author: these are targeted at different audiences. the entire first section ("why did we create this policy?") is targeted at people who aren't in the project. this section is targeted at new contributors, as a durable reference we can link to.
